mgr: ensure that all modules have started before advertising active mgr

ljflores opened the pull request:
When the mgr is restarted or failed over via `ceph mgr fail` or during an
upgrade, mgr modules sometimes take longer to start up (this includes
loading their class, commands, and module options, and being removed
from the `pending_modules` map structure). This startup delay can happen
due to a cluster's specific hardware or if a code bottleneck is triggered in
a module’s `serve()` function (each mgr module has a `serve()` function that
performs initialization tasks right when the module is loaded).

When this startup delay occurs, any mgr module command issued against the
cluster around the same time fails with error saying that the command is not
supported:
```
$ ceph mgr fail; ceph fs volume ls
Error ENOTSUP: Warning: due to ceph-mgr restart, some PG states may not be up to date
Module 'volumes' is not enabled/loaded (required by command 'fs volume ls'): use `ceph mgr module enable volumes` to enable it
```

We should try to lighten any bottlenecks in the mgr module `serve()`
functions wherever possible, but the root cause of this failure is that the
mgr sends a beacon to the mon too early, indicating that it is active before
the module loading has completed. Specifically, some of the mgr modules
have loaded their class but have not yet been deleted from the `pending_modules`
structure, indicating that they have not finished starting up.

This commit improves the criteria for sending the “active” beacon to the mon so
the mgr does not signal that it’s active too early. We do this through the following additions:

1. A new context `ActivePyModules::recheck_modules_start` that will be set if not all modules
   have finished startup.

2. A new function `ActivePyModules::check_all_modules_started()` that checks if modules are
   still pending startup; if all have started up (`pending_modules` is empty), then we send
   the beacon right away. But if some are still pending, we pass the beacon task on to the new
   recheck context `ActivePyModules::recheck_modules_start` so we know to send the beacon later.

3. Logic in ActivePyModules::start_one() that only gets triggered if the modules did not all finish
   startup the first time we checked. We know this is the case if the new recheck context
   `recheck_modules_start` was set from `nullptr`. The beacon is only sent once `pending_modules` is
   confirmed to be empty, which means that all the modules have started up and are ready to support commands.

4. Adjustment of when the booleans `initializing` and `initialized` are set. These booleans come into play in
   MgrStandby::send_beacon() when we check that the active mgr has been initialized (thus, it is available).
   We only send the beacon when this boolean is set. Currently, we set these booleans at the end of Mgr::init(),
   which means that it gets set early before `pending_modules` is clear. With this adjustment, the bools are set
   only after we check that all modules have started up. The send_beacon code is triggered on mgr failover AND on
   every Mgr::tick(), which occurs by default every two seconds. If we don’t adjust when these bools are set, we
   only fix the mgr failover part, but the mgr still sends the beacon too early via Mgr::tick(). Below is the relevant
   code from MgrStandby::send_beacon(), which is triggered in Mgr::background_init() AND in Mgr::tick():
```
  // Whether I think I am available (request MgrMonitor to set me
  // as available in the map)
  bool available = active_mgr != nullptr && active_mgr->is_initialized();

  auto addrs = available ? active_mgr->get_server_addrs() : entity_addrvec_t();
  dout(10) << "sending beacon as gid " << monc.get_global_id() << dendl;

```

At face value, this issue is indeterministically reproducible since it
can depend on environmental factors or specific cluster workloads.
However, I was able to deterministically reproduce it by injecting a
bottleneck into the balancer module:
```
diff --git a/src/pybind/mgr/balancer/module.py b/src/pybind/mgr/balancer/module.py
index https://github.com/ceph/ceph/commit/d12d69ffb01d973307b5eaa341f741811b6c4aa7..91c83fa8023 100644
--- a/src/pybind/mgr/balancer/module.py
+++ b/src/pybind/mgr/balancer/module.py
@@ -772,10 +772,10 @@ class Module(MgrModule):
                     self.update_pg_upmap_activity(plan)  # update pg activity in `balancer status detail`
                 self.optimizing = False
+                # causing a bottleneck
+                for i in range(0, 1000):
+                    for j in range (0, 1000):
+                        x = i + j
+                        self.log.debug("hitting the bottleneck in the balancer module")
             self.log.debug('Sleeping for %d', sleep_interval)
             self.event.wait(sleep_interval)
             self.event.clear()
```

Then, the error reproduces every time by running:
```
$ ./bin/ceph mgr fail; ./bin/ceph telemetry show
Error ENOTSUP: Warning: due to ceph-mgr restart, some PG states may not be up to date
Module 'telemetry' is not enabled/loaded (required by command 'telemetry show'): use `ceph mgr module enable telemetry` to enable it
```

With this fix, the active mgr is marked as "initialized" only after all
the modules have started up, and this error goes away. The command may
take a bit longer to execute depending on the extent of the delay.

### Part 2 -- Adding a "max expiration" mechanism + health warning to address extreme loading cases

During a mgr failover, the active mgr is marked available if:
      1. The mon has chosen a standby to be active
      2. The chosen active mgr has all of its modules initialized
    
Now that we've improved the criteria for sending the "active" beacon by enforcing it to retry initializing mgr modules, we need to account for extreme cases in which the modules are stuck loading for a very long time, or even indefinitely. In these extreme cases where the modules might never initialize, we don't want to delay sending the "active" beacon for too long. This can result in blocking other important mgr functionality, such as reporting PG availability in the health status. We want to avoid sending warnings about PGs being unknown in the health status when that's not ultimately the problem.

To account for an exceptionally long module loading time, I added a new configurable `mgr_module_load_max_expiration`. After a certain expiration (by default, 20 seconds), if the mgr modules still haven't finished initializing, the standby will then proceed to mark itself "available" and send the "active" beacon to the mon, which unblocks other critical mgr functionality.

For added clarity, a health error will be issued at this time, indicating which mgr modules got stuck initializing (See src/mgr/PyModuleRegistry.cc). The idea is to unblock the rest of the mgr's critical functionality while making it clear to Ceph operators that some modules are unusable.

This is what the health warning might look like:
```
$ ceph health detail

HEALTH_ERR 4 mgr modules have failed
[ERR] MGR_MODULE_ERROR: 14 mgr modules have failed
    Module 'rbd_support' has failed: Module failed to initialize.
    Module 'status' has failed: Module failed to initialize.
    Module 'telemetry' has failed: Module failed to initialize.
    Module 'volumes' has failed: Module failed to initialize.
```

Links to rendered documentation:
- https://ceph--63859.org.readthedocs.build/en/63859/mgr/administrator/
- https://ceph--63859.org.readthedocs.build/en/63859/rados/operations/health-checks/#mgr-module-error

Note: All of the new configurations added in this PR are marked as "dev" level, and are to be used only in testing scenarios or in out-of-the-norm troubleshooting scenarios. I have chosen not to document them for this reason.

Fixes: https://tracker.ceph.com/issues/71631
Signed-off-by: Laura Flores <lflores@ibm.com>

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

ljflores committed:
mgr: ensure that all modules have started before advertising active mgr

----------------- Explanation of Problem ----------------

When the mgr is restarted or failed over via `ceph mgr fail` or during an
upgrade, mgr modules sometimes take longer to start up (this includes
loading their class, commands, and module options, and being removed
from the `pending_modules` map structure). This startup delay can happen
due to a cluster's specific hardware or if a code bottleneck is triggered in
a module’s `serve()` function (each mgr module has a `serve()` function that
performs initialization tasks right when the module is loaded).

When this startup delay occurs, any mgr module command issued against the
cluster around the same time fails with error saying that the command is not
supported:
```
$ ceph mgr fail; ceph fs volume ls
Error ENOTSUP: Warning: due to ceph-mgr restart, some PG states may not be up to date
Module 'volumes' is not enabled/loaded (required by command 'fs volume ls'): use `ceph mgr module enable volumes` to enable it
```

We should try to lighten any bottlenecks in the mgr module `serve()`
functions wherever possible, but the root cause of this failure is that the
mgr sends a beacon to the mon too early, indicating that it is active before
the module loading has completed. Specifically, some of the mgr modules
have loaded their class but have not yet been deleted from the `pending_modules`
structure, indicating that they have not finished starting up.

--------------------- Explanation of Fix  --------------------

This commit improves the criteria for sending the “active” beacon to the mon so
the mgr does not signal that it’s active too early. We do this through the following additions:

1. A new context `ActivePyModules::recheck_modules_start` that will be set if not all modules
   have finished startup.

2. A new function `ActivePyModules::check_all_modules_started()` that checks if modules are
   still pending startup; if all have started up (`pending_modules` is empty), then we send
   the beacon right away. But if some are still pending, we pass the beacon task on to the new
   recheck context `ActivePyModules::recheck_modules_start` so we know to send the beacon later.

3. Logic in ActivePyModules::start_one() that only gets triggered if the modules did not all finish
   startup the first time we checked. We know this is the case if the new recheck context
   `recheck_modules_start` was set from `nullptr`. The beacon is only sent once `pending_modules` is
   confirmed to be empty, which means that all the modules have started up and are ready to support commands.

4. Adjustment of when the booleans `initializing` and `initialized` are set. These booleans come into play in
   MgrStandby::send_beacon() when we check that the active mgr has been initialized (thus, it is available).
   We only send the beacon when this boolean is set. Currently, we set these booleans at the end of Mgr::init(),
   which means that it gets set early before `pending_modules` is clear. With this adjustment, the bools are set
   only after we check that all modules have started up. The send_beacon code is triggered on mgr failover AND on
   every Mgr::tick(), which occurs by default every two seconds. If we don’t adjust when these bools are set, we
   only fix the mgr failover part, but the mgr still sends the beacon too early via Mgr::tick(). Below is the relevant
   code from MgrStandby::send_beacon(), which is triggered in Mgr::background_init() AND in Mgr::tick():
```
  // Whether I think I am available (request MgrMonitor to set me
  // as available in the map)
  bool available = active_mgr != nullptr && active_mgr->is_initialized();

  auto addrs = available ? active_mgr->get_server_addrs() : entity_addrvec_t();
  dout(10) << "sending beacon as gid " << monc.get_global_id() << dendl;

```

--------------------- Reproducing the Bug ----------------------

At face value, this issue is indeterministically reproducible since it
can depend on environmental factors or specific cluster workloads.
However, I was able to deterministically reproduce it by injecting a
bottleneck into the balancer module:
```
diff --git a/src/pybind/mgr/balancer/module.py b/src/pybind/mgr/balancer/module.py
index d12d69f..91c83fa8023 100644
--- a/src/pybind/mgr/balancer/module.py
+++ b/src/pybind/mgr/balancer/module.py
@@ -772,10 +772,10 @@ class Module(MgrModule):
                     self.update_pg_upmap_activity(plan)  # update pg activity in `balancer status detail`
                 self.optimizing = False
+                # causing a bottleneck
+                for i in range(0, 1000):
+                    for j in range (0, 1000):
+                        x = i + j
+                        self.log.debug("hitting the bottleneck in the balancer module")
             self.log.debug('Sleeping for %d', sleep_interval)
             self.event.wait(sleep_interval)
             self.event.clear()
```

Then, the error reproduces every time by running:
```
$ ./bin/ceph mgr fail; ./bin/ceph telemetry show
Error ENOTSUP: Warning: due to ceph-mgr restart, some PG states may not be up to date
Module 'telemetry' is not enabled/loaded (required by command 'telemetry show'): use `ceph mgr module enable telemetry` to enable it
```

With this fix, the active mgr is marked as "initialized" only after all
the modules have started up, and this error goes away. The command may
take a bit longer to execute depending on the extent of the delay.

---------------------- Integration Testing ---------------------

This commit adds a dev-only config that can inject a longer
loading time into the mgr module loading sequence so we can
simulate this scenario in a test.

The config is 0 ms by default since we do not add any delay
outside of testing scenarios. The config can be adjusted
with the following command:
  `ceph config set mgr mgr_module_load_delay <ms>`

A second dev-only config also allows you to specify which
module you want to be delayed in loading time. You may change
this with the following command:
  `ceph config set mgr mgr_module_load_delay_name <module name>`

The workunit added here tests a simulated slow loading module
scenario to ensure that this case is properly handled.

--------------------- Documentation --------------------

The new documentation describes the three existing mgr states so Ceph
operators can better interpret their Ceph status output.

Fixes: https://tracker.ceph.com/issues/71631
Signed-off-by: Laura Flores <lflores@ibm.com>

ljflores commented:
jenkins test windows

athanatos reviewed (approved):
LGTM, but someone more familiar with the manager should probably have a look as well.

ljflores commented:
Changed the commit description; code is the same.

batrick reviewed (changes requested):
I haven't studied this code in detail yet. I have a question on the overall approach here: why not just have the `ceph-mgr` sit on `MMgrCommand`s and retry all tabled commands whenever a module becomes active. If a module never becomes active after some timeout, return `-ENOSYS` (**not** `-ENOTSUP`).

That way the ceph-mgr is available immediately for modules which load quickly and others just delay the command delivery until a module that understands the command successfully loads.

athanatos commented:
> I haven't studied this code in detail yet. I have a question on the overall approach here: why not just have the `ceph-mgr` sit on `MMgrCommand`s and retry all tabled commands whenever a module becomes active. If a module never becomes active after some timeout, return `-ENOSYS` (**not** `-ENOTSUP`).
> 
> That way the ceph-mgr is available immediately for modules which load quickly and others just delay the command delivery until a module that understands the command successfully loads.

I don't see any real advantage to allowing the manager to be "active" without the modules being loaded -- seems a lot more complicated for little advantage.  Yes, it allows a slightly faster startup, but that hardly seems critical on the manager.

ljflores commented:
> My thought was that you'd keep a vector of MMgrCommands to retry whenever a module becomes active. There would be no need to sleep the messenger thread.

I still don’t think this would offer any significant benefit. Mgr modules load sequentially and depend on each other, so we can already infer a module’s availability based on whether the ones before it have loaded. Retrying tabled commands individually adds extra complexity, especially when we know every enabled module is expected to be loaded and functional in the cluster.

> The larger problem is that the ceph-mgr doesn't actually decide when it's active. The monitors tell the standby it's active.

I disagree with you on this. The monitor AND the active mgr it chooses both have a hand in determining the mgr’s availability. In `MgrStandby::send_beacon()`, the “available” status is set based on two conditions:
1. The mon must choose a new active mgr → decision made by mon
2. The new active mgr must be initialized. → decision made by the chosen active mgr

The chosen active mgr ultimately tells the mon when it’s done initializing, and it is actually an expectation that the python modules be loaded per the comment written under the “if (available)” condition.

The code that I’m referencing is here:
```
void MgrStandby::send_beacon()
{
 …
  // Whether I think I am available (request MgrMonitor to set me
  // as available in the map)
  bool available = active_mgr != nullptr && active_mgr->is_initialized();
…
if (available) {
    if (!available_in_map) {
      // We are informing the mon that we are done initializing: inform
      // it of our command set.  This has to happen after init() because
      // it needs the python modules to have loaded.
      std::vector<MonCommand> commands = mgr_commands;
      std::vector<MonCommand> py_commands = py_module_registry.get_commands();
      commands.insert(commands.end(), py_commands.begin(), py_commands.end());
      if (monc.monmap.min_mon_release < ceph_release_t::quincy) {
        dout(10) << " stripping out positional=false quincy-ism" << dendl;
        for (auto& i : commands) {
          boost::replace_all(i.cmdstring, ",positional=false", "");
        }
      }
      m->set_command_descs(commands);
      dout(4) << "going active, including " << m->get_command_descs().size()
              << " commands in beacon" << dendl;
    }

    m->set_services(active_mgr->get_services());
  }
```

> If the ceph-mgr doesn't immediately start sending beacons then the mons will assume it's dead and replace it.

That’s not quite what’s happening here. In my solution, the mgr does keep sending beacons, just not the final “active” beacon until all modules have started. While it's waiting for initialization to complete, it sends a beacon indicating that it’s “active (starting)”. So the mon is being notified continuously, and doesn’t consider the mgr dead. Even if a module is stalled loading forever, the mon doesn’t replace the chosen active mgr. It waits for the mgr to signal that initialization is complete.

See this comment for a clear demonstration of that: https://github.com/ceph/ceph/pull/63859#issuecomment-2964092195

> This seems to be the reason why the "available" flag was invented but its meaning is controversial and recently discussed in e.g.

> #51169 (comment)

> I'd argue the "available" flag is half-baked and we keep churning on this continually. I think a complete solution requires adding recovery states to the ceph-mgr starting with something like "STATE_LOADING_MODULES".

Looking at the MgrStandby::send_beacon code, it seems straightforward to me that the “available” condition is set by 1) the mon choosing a new active_mgr and 2) the new active mgr signaling initialization. Introducing module-by-module availability would again break these expectations and make communication with the mon and mgr command handler much more difficult. That said, I'm open to understanding more about why the condition might be half-baked and how we can make it clearer.

> The ceph-mgr eventually should tell the mons when it's ready to switch to active in the same way that the MDS do.

Can you please elaborate on the active sequence in the MDS and how it differs from the mgr?

ljflores commented:
> I'm not actually sure this is a big deal. I haven't actually seen evidence that loading modules takes a long time, merely that there's a possible race condition if we send the beacon before it's actually happened. Perhaps we try this solution and see if the delay is actually a problem?

I know you and I have gone over this scenario many times, but I want to point out the original example that motivated this fix, which is documented in https://bugzilla.redhat.com/show_bug.cgi?id=2314146. In this report, the QE team was testing upgrades from Reef/Quincy to Squid, and their scripts were erroring out as soon as the mgrs upgraded to Squid. We had not seen this race condition occur before Squid, and after digging into it, I found that it is likely due to us gradually adding more load on the various `serve()` functions that are triggered for every enabled module upon initialization.

One might suggest optimizing these `serve()` functions. Sometimes that is the right call; in a related issue (https://tracker.ceph.com/issues/68657), an extra function was added to the balancer module's `serve()` function in Squid, which caused the module to load more slowly and exacerbate this error. We fixed the bottleneck since it wasn't necessary to the balancer's functionality. However, what happens when we want to add extra tasks to a module's `serve()` function out of necessity? That's a very valid scenario that should be accounted for in the module loading sequence.

In the example from https://bugzilla.redhat.com/show_bug.cgi?id=2314146, the QE team is still experiencing the error after we fixed the bottleneck in the balancer, although to a lesser extent. I experimented with disabling cephadm to remove it from the loading sequence, and then the error "went away". This indicates that cephadm might have a heavy `serve()` function as well, and sure enough there were some additions to cephadm's `serve()` function that got added to Squid that may be related. However, even if we attempt to optimize that, those additions might be necessary to cephadm's functionality. Also, that doesn't address the root cause, which is that the mgr is declaring availability too early.

Ultimately, we likely didn't see this problem before Squid since we got lucky. But, with the introduction of the finisher thread framework, which changed the module loading sequence to an async background task, and by us gradually adding more tasks to the `serve()` functions, it was only a matter of time before this race occurred. This fix will be most beneficial in scenarios where we are running mgr commands against mgrs that are in the middle of upgrading, failing over, or otherwise initializing.

rzarzynski reviewed (commented):
I think I like this solution. I'm close to approving with the only reason for not doing it right now is a paranoid, unproven concern about atomicity of the flags management. 

I agree that optimizing the mgr's boot time is rather not a concern.

ljflores commented on src/mgr/Mgr.cc:
```diff
@@ -171,7 +171,12 @@ void Mgr::background_init(Context *completion)
+    py_module_registry->check_all_modules_started(
+	new LambdaContext([this, completion](int){
+	initializing = false;
+	initialized = true;
```
Yes, you understand correctly!

mchangir commented:
> > I would also ask @idryomov to chime in here as I think this PR is doing what he previously asked we do.
> 
> Sure, I'll request Ilya as a reviewer.
> 
> > No sorry, I didn't mean that the ceph-mgr should indicate via a state change for each individual module. It would just be a recovery state. The "available" flag approximates that to a degree. What I don't like is that operators won't understand why the mgr may not be working when it's clearly "active". Perhaps that's something that can be included in this PR: the mon's can log a health warning in MgrMonitor::encode_pending that the mgr is active but not yet available. (This is normal as we log similar warnings when the MDS is in a degraded state.)
> 
> FWIW, the ceph status does already indicate that the new active mgr is still starting. This snippet was taken during the same "initialization sequence delay via balancer bottleneck" that I illustrated in [#63859 (comment)](https://github.com/ceph/ceph/pull/63859#issuecomment-2964092195):
> 
> ```
> $ ./bin/ceph -s
>   cluster:
>     id:     b150f540-745a-460c-a566-376b28b95ac3
>     health: HEALTH_OK
>  
>   services:
>     mon: 3 daemons, quorum a,b,c (age 47m) [leader: a]
>     mgr: x(active, starting, since 3s)
>     mds: 1/1 daemons up, 2 standby
>     osd: 4 osds: 4 up (since 47m), 4 in (since 47m)
>  
>   data:
>     volumes: 1/1 healthy
>     pools:   4 pools, 177 pgs
>     objects: 24 objects, 451 KiB
>     usage:   4.0 GiB used, 400 GiB / 404 GiB avail
>     pgs:     177 active+clean
> ```
> 
> The MgrStandby code defines the three possible states: "standby", "active", and "active (starting)":
> 
> ```
> std::string MgrStandby::state_str()
> {
>   if (active_mgr == nullptr) {
>     return "standby";
>   } else if (active_mgr->is_initialized()) {
>     return "active";
>   } else {
>     return "active (starting)";
>   }
> ```
> 
> `active (starting)` is synonymous with "pending", and should indicate to an operator that the mgr is still starting up, but perhaps this can be better documented. A quick grep of "active (starting)" didn't reveal anything in the documentation, so we could place a section there explaining what each of these three mgr states means. I can also look into adding a cluster warning, but it's worth noting that in most field examples of module loading delay, the delay has historically only been a second or two, so any new warning would only appear for short time. Still, I could foresee a case where the loading sequence gets stuck (this has not happened as of yet in the field), but we should probably warn about that.
> 
> > In any case, any necessary redesign with recovery states can be done in the future.
> 
> Sure, it would be helpful to have a design proposal in the form of an Enhancement tracker so all the criteria and intended benefits are clearly stated.

Would changing the state description from `active (starting)` to `active (loading modules)` help any bit ? Or maybe don't mention the `active` part ... just mention state as `loading modules`.
What else other than loading modules does the mgr do after declaring itself as `active` ?

rzarzynski commented on src/mgr/Mgr.cc:
```diff
@@ -171,7 +171,12 @@ void Mgr::background_init(Context *completion)
-    completion->complete(0);
+    py_module_registry->check_all_modules_started(
+	new LambdaContext([this, completion](int){
+	initializing = false;
```
Yes, within the `LambdaContext` but you need to watch about locking-already-locked `lock` from the `completion`.

ljflores replied:
Added a lock around "initialized" and "initializing". Let me know if this looks okay to you.

ljflores committed:
mgr, common, qa, doc: issue health error after max expiration is exceeded

----------------- Enhancement to the Original Fix -----------------

During a mgr failover, the active mgr is marked available if:
  1. The mon has chosen a standby to be active
  2. The chosen active mgr has all of its modules initialized

Now that we've improved the criteria for sending the "active" beacon
by enforcing it to retry initializing mgr modules, we need to account
for extreme cases in which the modules are stuck loading for a very long
time, or even indefinitely. In these extreme cases where the modules might
never initialize, we don't want to delay sending the "active" beacon for
too long. This can result in blocking other important mgr functionality,
such as reporting PG availability in the health status. We want
to avoid sending warnings about PGs being unknown in the health status when
that's not ultimately the problem.

To account for an exeptionally long module loading time, I added a new
configurable `mgr_module_load_expiration`. If we exceed this maximum amount
of time (in ms) allotted for the active mgr to load the mgr modules before declaring
availability, the standby will then proceed to mark itself "available" and
send the "active" beacon to the mon and unblock other critical mgr functionality.

If this happens, a health error will be issued at this time, indicating
which mgr modules got stuck initializing (See src/mgr/PyModuleRegistry.cc). The
idea is to unblock the rest of the mgr's critical functionality while making it
clear to Ceph operators that some modules are unusable.

--------------------- Integration Testing --------------------

The workunit was rewritten so it tests for these scenarios:

1. Normal module loading behavior (no health error should be issued)
2. Acceptable delay in module loading behavior (no health error should be
   issued)
3. Unacceptable delay in module loading behavior (a health error should be
   issued)

--------------------- Documentation --------------------

This documentation explains the "Module failed to initialize"
cluster error.

Users are advised to try failing over
the mgr to reboot the module initialization process,
then if the error persists, file a bug report. I decided
to write it this way instead of providing more complex
debugging tips such as advising to disable some mgr modules
since every case will be different depending on which modules
failed to initialize.

In the bug report, developers can ask for the health detail
output to narrow down which module is causing a bottleneck,
and then ask the user to try disabling certain modules until
the mgr is able to fully initialize.

Fixes: https://tracker.ceph.com/issues/71631
Signed-off-by: Laura Flores <lflores@ibm.com>

ljflores commented:
> /a/vshankar-2025-07-21_12:33:32-fs-wip-khiremat-testing-20250628.025402-tentacle-debug-testing-default-smithi/8399441

@vshankar I checked out this job, and I don't see the telltale error in the mgr log:
```
Error ENOTSUP: Warning: due to ceph-mgr restart, some PG states may not be up to date
Module 'volumes' is not enabled/loaded (required by command 'fs volume ls'): use `ceph mgr module enable volumes` to enable it
```
I thought it could be possible that the error is the same and presenting in a different way, but then I noticed this traceback earlier in the same mgr log. Could the root cause be this?
```
2025-07-21T13:27:00.629+0000 7f8f48e3b640 20 client.7342 _lookup have 0x1.head["bad-path" ref=2 ino=nil csg=-1] from mds.-1 ttl 0.000000 seq 0
2025-07-21T13:27:00.629+0000 7f8f48e3b640 10 client.7342 _lookup concluded ENOENT locally for 0x1.head(faked_ino=0 nref=11 ll_ref=1 cap_refs={} open={} mode=41777 size=0/0 nlink=1 btime=2025-07-21T13:26:24.528962+0000 mtime=2025-07-21T13:26:24.528962+0000 ctime=2025-07-21T13:26:36.310257+0000 change_attr=1 caps=pAsLsXsFs(0=pAsLsXsFs) COMPLETE has_dir_layout 0x55bea26d1e00) dn 'bad-path'
2025-07-21T13:27:00.629+0000 7f8f48e3b640 10 client.7342 path_walk: rc=-2 result=walk(diri=0x1.head name="bad-path" ENOENT)
2025-07-21T13:27:00.629+0000 7f8f48e3b640 20 client.7342 put_inode on 0x1.head(faked_ino=0 nref=11 ll_ref=1 cap_refs={} open={} mode=41777 size=0/0 nlink=1 btime=2025-07-21T13:26:24.528962+0000 mtime=2025-07-21T13:26:24.528962+0000 ctime=2025-07-21T13:26:36.310257+0000 change_attr=1 caps=pAsLsXsFs(0=pAsLsXsFs) COMPLETE has_dir_layout 0x55bea26d1e00) n = 1
2025-07-21T13:27:00.629+0000 7f8f48e3b640 20 client.7342 put_inode on 0x1.head(faked_ino=0 nref=11 ll_ref=1 cap_refs={} open={} mode=41777 size=0/0 nlink=1 btime=2025-07-21T13:26:24.528962+0000 mtime=2025-07-21T13:26:24.528962+0000 ctime=2025-07-21T13:26:36.310257+0000 change_attr=1 caps=pAsLsXsFs(0=pAsLsXsFs) COMPLETE has_dir_layout 0x55bea26d1e00) n = 1
2025-07-21T13:27:00.629+0000 7f8f48e3b640 20 client.7342 put_inode on 0x1.head(faked_ino=0 nref=11 ll_ref=1 cap_refs={} open={} mode=41777 size=0/0 nlink=1 btime=2025-07-21T13:26:24.528962+0000 mtime=2025-07-21T13:26:24.528962+0000 ctime=2025-07-21T13:26:36.310257+0000 change_attr=1 caps=pAsLsXsFs(0=pAsLsXsFs) COMPLETE has_dir_layout 0x55bea26d1e00) n = 1
2025-07-21T13:27:00.630+0000 7f8f48e3b640  0 [snap_schedule DEBUG mgr_util] [put] connection: <mgr_util.CephfsConnectionPool.Connection object at 0x7f8f445f8190> usage: 1
2025-07-21T13:27:00.630+0000 7f8f48e3b640  0 [snap_schedule ERROR snap_schedule.fs.schedule_client] create_scheduled_snapshot raised an exception:
2025-07-21T13:27:00.630+0000 7f8f48e3b640  0 [snap_schedule ERROR snap_schedule.fs.schedule_client] Traceback (most recent call last):
  File "/usr/share/ceph/mgr/snap_schedule/fs/schedule_client.py", line 355, in create_scheduled_snapshot
    fs_handle.mkdir(snap_name, 0o755)
  File "cephfs.pyx", line 1123, in cephfs.LibCephFS.mkdir
cephfs.ObjectNotFound: error in mkdir /bad-path/.snap/scheduled-2025-07-21-13_27_00_UTC: No such file or directory [Errno 2]

2025-07-21T13:27:00.630+0000 7f8f48e3b640  0 [snap_schedule DEBUG snap_schedule.fs.schedule_client] path /bad-path is probably missing or wrong; remember to strip off the mount point path prefix to provide the correct path
```

anthonyeleven reviewed (commented):
Nice enhancements.  I left various fussy docs suggestions.

ljflores commented on doc/mgr/administrator.rst:
```diff
@@ -31,6 +31,52 @@ which should now include a mgr status line::
 
     mgr active: $name
 
+Interpreting Ceph-Mgr Statuses
```
Thanks! I'll address all of these once everyone else has had a chance to review the changes so I'm not force-pushing a million times.

batrick commented on src/mgr/Mgr.cc:
```diff
@@ -756,6 +763,23 @@ bool Mgr::got_mgr_map(const MgrMap& m)
+{
+  std::lock_guard l(lock);
+  if (!is_initialized()) {
+    initialization_retries++;
```
I don't really like the connection between the beacon period and this "retry" counter. The code is not really retrying to load the modules either. I would prefer a simple `ceph::coarse_mono_time Mgr::initialization_started`. Then force active when we exceed some threshold `mgr_module_load_expiration` (perhaps).

batrick commented on src/mgr/Mgr.cc:
```diff
@@ -756,6 +763,23 @@ bool Mgr::got_mgr_map(const MgrMap& m)
   return false;
 }
 
+bool Mgr::exceeded_initialization_retires() const
```
spelling

batrick commented on src/mgr/MgrStandby.cc:
```diff
@@ -277,8 +277,26 @@ void MgrStandby::send_beacon()
+  // are unusuable.
+  bool available = false;
+  if (active_mgr != nullptr) {
+    available = active_mgr->is_initialized() || active_mgr->exceeded_initialization_retires();
```
I wonder if there is any inconsistency or potential for problems that we're marking the Mgr `available` but internally it's still `initializing`?

ljflores replied:
Ah, do you mean that `initialized` and `initializing` booleans should be set to "True" and "False" respectively at this point? If so, I agree they should be set if we are declaring the mgr active. Let me know.

batrick commented on src/mgr/ActivePyModules.cc:
```diff
@@ -538,6 +538,14 @@ void ActivePyModules::start_one(PyModuleRef py_module)
+              << "' by " << delay << "..." << dendl;
+      std::this_thread::sleep_for(delay);
+    }
     int r = active_module->load(this);
```
Something to consider which may help with testing: have a new mgr asok command that dumps module state for  `pending_modules` and `modules`.

batrick commented on src/mgr/ActivePyModules.cc:
```diff
@@ -538,6 +538,14 @@ void ActivePyModules::start_one(PyModuleRef py_module)
   // Send all python calls down a Finisher to avoid blocking
   // C++ code, and avoid any potential lock cycles.
   finisher.queue(new LambdaContext([this, active_module, name](int) {
```
So an issue I see here is that this module failing to load is presumably starving this finisher. We now have a per-module finisher so this may not actually block the completion of most work but, I just checked, we have code like `DaemonServer::handle_osd_perf_metric_query_updated` which does queue contexts on this finisher. If a module is legit stuck loading indefinitely then those finishers will never run. Correct? We also can see that we have a pipeline of modules loading and one becoming stuck blocks all subsequent modules, no?

So that means we need a way to terminate the module load (probably not trivial) or launch a thread for each module load to avoid blocking this Finisher thread. My preference would be to just launch a thread which loads the module then exits.

batrick replied:
Well, looking at this code again. I see that there is `class ActivePyModule : public PyModuleRunner` which is created just before this code above (`active_module`). That `PyModuleRunner` actually creates a thread already just for calling `->serve()` in the module. Well, we could just have that thread first `load` and only then `serve`. If you attach a context to the PyModuleRunner then you can have it complete the context after `load` completes. That would allow you to execute this lambda above.

Anyway, I think that's the most elegant way to resolve this.

ljflores replied:
ACK, I created a ticket here: https://tracker.ceph.com/issues/73060

batrick reviewed (changes requested):
Mostly minor things that are easily addressed. This is good work.

My other concern however is that a pending module load continues even after the mgr forcibly goes `available`. That should be made clear in the docs / warning that the module load may still complete.

ljflores committed:
mgr, qa: add `pending_modules` to asock command

Now, the command `ceph tell mgr mgr_status` will show a
"pending_modules" field. This is another way for Ceph operators
to check which modules haven't been initalized yet (in addition
to the health error).

This command was also added to testing scenarios in the workunit.

Fixes: https://tracker.ceph.com/issues/71631
Signed-off-by: Laura Flores <lflores@ibm.com>

ljflores committed:
mgr, qa: clarify module checks in DaemonServer

The current check groups modules not being
enabled with failing to initialize. In this commit,
we reorder the checks:

1: Screen for a module being enabled. If it's not,
   issue an EOPNOTSUPP with instructions on how
   to enable it.

2. Screen for if a module is active. If a module
   is enabled, then the cluster expects it to
   be active to support commands. If the module
   took too long to initialize though, we will
   catch this and issue an ETIMEDOUT error with
   a link for troubleshooting.

Now, these two separate issues are not grouped
together, and they are checked in the right order.

Fixes: https://tracker.ceph.com/issues/71631
Signed-off-by: Laura Flores <lflores@ibm.com>

anthonyeleven commented on doc/mgr/administrator.rst:
```diff
@@ -31,6 +31,52 @@ which should now include a mgr status line::
+
+2. **active (starting)**
+
+   This manager daemon has been chosen to be ``active``, but it is not done initializing.
```
Nit: There’s a bit of a trend to capitalize Manager and Monitor. Here and below.

anthonyeleven commented on doc/mgr/administrator.rst:
```diff
@@ -31,6 +31,52 @@ which should now include a mgr status line::
+3. **standby**
+
+   This manager daemon is not currently receiving or executing commands, but it is ready to
+   take over if the current active manager becomes unavailable. An administrator may also
```
Suggest:

s/also//

anthonyeleven commented on doc/mgr/administrator.rst:
```diff
@@ -31,6 +31,52 @@ which should now include a mgr status line::
+   manually promote a standby to become active via ``ceph mgr fail`` if desired. All other
+   manager daemons which are not ``active`` or ``active (starting)`` will be in this state.
+
+Each of these states are visible in the output of the ``ceph status`` command. For example:
```
Maybe mention `ceph mgr stat` too?

anthonyeleven commented on doc/rados/operations/health-checks.rst:
```diff
@@ -310,6 +310,31 @@ However, if you believe the error is transient, you may restart your manager
+**Module failed to initialize**
+
+If the ``ceph health detail`` looks something like this, it means that some
+modules took too long to initialize post a mgr failover, and are unable to process
```
Suggest:

s/this/the below/
s/post a mgr/after a Manager/

anthonyeleven commented on doc/rados/operations/health-checks.rst:
```diff
@@ -310,6 +310,31 @@ However, if you believe the error is transient, you may restart your manager
+       Module 'telemetry' has failed: Module failed to initialize.
+       Module 'volumes' has failed: Module failed to initialize.
+
+To troubleshoot, you may try running ``ceph mgr fail`` to reboot
```
Suggest:

s/you may try running/run/

anthonyeleven commented on doc/rados/operations/health-checks.rst:
```diff
@@ -310,6 +310,31 @@ However, if you believe the error is transient, you may restart your manager
+To troubleshoot, you may try running ``ceph mgr fail`` to reboot
+module initialization.
+
+Note that the health error may disappear on its own since modules
```
Suggest s/disappear/clear/

anthonyeleven commented on qa/workunits/mgr/test_mgr_module_loading_time.sh:
```diff
@@ -0,0 +1,215 @@
+if [[ "$mgr_status" == "$expected" ]]; then
+    echo "PASS: No modules are pending."
+else
+    echo "FAIL: Some modules are pending when there shouldn't be."
```
I think this should be either

be any

or

when they shouldn’t be.

anthonyeleven commented on src/common/options/mgr.yaml.in:
```diff
@@ -160,6 +160,39 @@ options:
+  default: 0
+  desc: Number of milliseconds for Manager modules to delay loading. For testing purposes only.
+  long_desc: Number of milliseconds for Manager modules to delay loading. For testing purposes only.
+   Related to `mgr_module_load_delay_name`.
```
Suggest making this a structured `see_also:`

anthonyeleven reviewed (commented):
Docs lgtm with a number of minor suggestions.

batrick reviewed (approved):
Good work!

ljflores commented:
> Nice detailed description.
> 
> > beacon to the mon so the mgr does not signal
> 
> Here, should “mon” be “mgr”?

No, the beacon is sent from the mgr to the mon.

NitzanMordhai reviewed (approved):
I reviewed all the changes, and it looks good!

ljflores committed:
qa/workunits/mgr: account for nvmeof module being "always-on"

Post the merge of this: https://github.com/ceph/ceph/pull/67641

Fixes: https://tracker.ceph.com/issues/71631
Signed-off-by: Laura Flores <lflores@ibm.com>

NitzanMordhai reviewed (approved):
LGTM
